### ISSUE TITLE
Manifold and Volume for groups / compoennts mixing faces and other groups

### ISSUE THREAD

**Issue by Fredosixx** [SketchUp/api-issue-tracker#698]
In complement to Issue #692, there is another limitation, which is both for the API and Sketchup GUI.

**When a group (or component instance) is composed of a set of faces forming a manifold shape, but also includes other groups or component instances**, it is not considered manifold and thus it is not shown as Manifold (with indication of volume) in the Entity Info panel.

It seems to me that the check on manifold and the calculation of volumes can be based **only on faces included in the group**, ignoring anything else (currently only non-geometry elements like guides, 2D-text, dimensions are ignored).

The typical case where I see problems is when you have a solid group where you put small 3d-text labels on top of each face, each label being a separated group within the solid group.

**Comment by thomthom** [comment]
> It seems to me that the check on manifold and the calculation of volume can be based only on faces included in the group,

Correct, SketchUp currently considers an instance manifold if all edges connect to exactly two faces and there are no sub-instances.

Can you elaborate on what this feature request it? (I assume it's a feature request?) For `.manifold?` to return true when the whole of a nested hierarchy is "solid"? For Solid tools to operator on that as well? Is this more than just an API request?

**Comment by Fredosixx** [comment]
This is a request for the API. However, I think it should laso go to the GUI for sake of consistency.

**_A group or component instance if manifold if the faces it contains form a solid shape REGARDLESS of any other non-face elements it contains._** (whether other groups, component instances, images, or even standalone edges).

With the above definition, I can access to the volume of the solid formed by the shapes (instead of having my own custom method)

I understand that it may have implications for STL export where it should may be use a more restrictive definition of manifold. But for the API an the GUI, that good information.

**Comment by DanRathbun** [comment]
What would you propose happen to the stray primitives during the various boolean operations ?

**Comment by Fredosixx** [comment]
With the new definition of manifold, the boolean operations would act on the faces of the 2 groups, ignoring the other groups. It does not need to cascade down.

**Comment by DanRathbun** [comment]
Okay, so basically **ignore** what is should to do the boolean job and report proper volumes.

**Comment by thomthom** [comment]
I don't think we can alter the existing method. Either a new method or an optional argument to opt in to alternate behaviour.

**Comment by DanRathbun** [comment]
> ... or an optional argument to opt in to alternate behaviour.

Opt-in is somewhat "clunky" IMO. I don't really think for example, that the "[break edges](https://ruby.sketchup.com/Sketchup.html#break_edges=-class_method)" opt in is used much.

> Either a new method ...

**New methods** that accept an array of faces might be added to the [`Sketchup::Entities`](https://ruby.sketchup.com/Sketchup/Entities.html) class:

* [`#manifold_form?`](faces) - Boolean if the faces passed would form a manifold solid (in the absence of other objects.)
The `faces` default would be an array of **all** faces in the collection.

* [`#volume_of(transform, faces)`]() - Volume of the "virtual" solid formed by the `faces` array applying the passed `transform`.
The `transform` default would be `IDENTITY`.
The `faces` default would be an array of **all** faces in the collection.
